File: TP1/Partie2_Calculer_Pi/pi_mpi.py
from mpi4py import MPI
import numpy as np
import random
import time

N_POINTS = 100000000

def estimate_pi(num_points, seed):
    random.seed(seed)
    count_in_circle = sum(1 for _ in range(num_points) 
                          if (random.uniform(-1, 1) ** 2 + random.uniform(-1, 1) ** 2) <= 1)
    return count_in_circle

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

points_per_proc = N_POINTS // size
start_time = time.time()

local_count = estimate_pi(points_per_proc, rank)
total_count = comm.reduce(local_count, op=MPI.SUM, root=0)

if rank == 0:
    pi = 4.0 * total_count / N_POINTS
    end_time = time.time()
    print(f"Estimation de Pi = {pi}")
    print(f"Temps d'exécution MPI avec {size} processus : {end_time - start_time:.6f} secondes")

File: TP1/Partie2_Calculer_Pi/pi_mpi.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

#define N_POINTS 100000000  // Nombre total de points

double estimate_pi(int num_points, int rank) {
    int count_in_circle = 0;
    unsigned int seed = rank + 1; // Seed unique pour chaque processus

    for (int i = 0; i < num_points; i++) {
        double x = (rand_r(&seed) / (double)RAND_MAX) * 2.0 - 1.0;
        double y = (rand_r(&seed) / (double)RAND_MAX) * 2.0 - 1.0;
        if (x * x + y * y <= 1.0) {
            count_in_circle++;
        }
    }
    return (double)count_in_circle;
}

int main(int argc, char *argv[]) {
    int rank, size;
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    int points_per_proc = N_POINTS / size;
    double start_time = MPI_Wtime();

    double local_count = estimate_pi(points_per_proc, rank);
    double total_count;

    MPI_Reduce(&local_count, &total_count, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

    if (rank == 0) {
        double pi = 4.0 * total_count / N_POINTS;
        double end_time = MPI_Wtime();
        printf("Estimation de Pi = %f\n", pi);
        printf("Temps d'exécution MPI avec %d processus : %f secondes\n", size, end_time - start_time);
    }

    MPI_Finalize();
    return 0;
}
